File: 백준/Gold/7576. 토마토/토마토.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	static int[][] matrix, visit;
	static int m,n, result;
	static int[] dx={0,0,-1,1}, dy={1,-1,0,0};
	static Queue<Node> q;
	public static void main(String[] args) throws Exception {
		st = new StringTokenizer(br.readLine());
		m = Integer.parseInt(st.nextToken()); //열
		n = Integer.parseInt(st.nextToken()); //행
		matrix = new int[n][m];
		visit = new int[n][m];
		q = new LinkedList<>();
		for (int i=0; i<n; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j=0; j<m; j++) {
				matrix[i][j] = Integer.parseInt(st.nextToken());
				if (matrix[i][j]==1) {
					q.add(new Node(i,j));
					visit[i][j]=1;
				}
				else if (matrix[i][j]==-1) {
					visit[i][j]=1;
				}
			}
		}
		result = -1;
		bfs();
		for (int r=0; r<n; r++) {
			for (int c=0; c<m; c++) {
				if (visit[r][c]==0) {
					result = -1;
					break;
				}
			}
		}
		System.out.println(result);
	}
	
	public static void bfs() {
		while (!q.isEmpty()) {
			int size = q.size();
			result++;
			for (int i=0; i<size; i++) {
				Node now = q.poll();
				for (int j=0; j<4; j++) {
					int nx = now.x+dx[j];
					int ny = now.y+dy[j];
					if (nx<0 || nx>=n || ny<0 || ny>=m)
						continue;
					if (matrix[nx][ny]==0 && visit[nx][ny]==0) {
						visit[nx][ny]=1;
						q.add(new Node(nx,ny));
					}
				}
			}
		}
	}
	
	public static class Node{
		int x;
		int y;
		public Node(int x, int y) {
			this.x = x;
			this.y = y;
		}	
	}
}

File: 백준/Gold/7576. 토마토/토마토.py
from collections import deque
dx,dy = [1,-1,0,0], [0,0,1,-1]

def bfs():
    q = deque()
    for i in range(n):
        for j in range(m):
            if matrix[i][j]==1:
                q.append((i,j))
    cnt = -1
    while q:
        size = len(q)
        cnt+=1
        for i in range(size):
            now = q.popleft()
            for k in range(4):
                nx = now[0]+dx[k]
                ny = now[1]+dy[k]
                if nx<0 or nx>=n or ny<0 or ny>=m:
                    continue
                if matrix[nx][ny]==0:
                    matrix[nx][ny]=1
                    q.append((nx,ny))
    for i in range(n):
        for j in range(m):
            if matrix[i][j]==0:
                return -1
    return cnt

m,n = map(int, input().split())
matrix = [list(map(int, input().split())) for _ in range (n)]
result = 0
print(bfs())